### DISCUSSION TITLE
What are the best practices when configuring the SDK to be optional in Rails?

### DISCUSSION THREAD

**Discussion by murshed-panorama** [open-telemetry/opentelemetry-ruby#1605] [Q&A] [answered]
Hello. I am configuring the SDK for a Rails application.

I would like to only enable instrumentation when an environment variable `OTEL_INSTRUMENTATION_ENABLED` is set. This may be true in some environments like staging, but not in local development, for example.

I have a configuration file in `my-app/config/initializers/opentelemetry.rb` that looks similar to this:

```ruby
require "opentelemetry/sdk"
require "opentelemetry/instrumentation/all"
require "opentelemetry-exporter-otlp"

Rails.application.configure do
  otel_enabled = ENV.fetch("OTEL_INSTRUMENTATION_ENABLED", false)
  OpenTelemetry::SDK.configure do |c|
    if otel_enabled
      Rails.logger.info("Configuring OpenTelemetry SDK")
      c.service_name = "MY_APP"
      c.use_all
    end

    # Traces will automatically be sent to endpoint specified
    # in the OTEL_EXPORTER_OTLP_ENDPOINT environment variable
  end

  if otel_enabled
    MyAppTracer = OpenTelemetry.tracer_provider.tracer("MY_APP_TRACER")
  end
end
```

This should prevent automatic instrumentation, but do I need additional checks when adding manual instrumentation?

For example, in one of my job's functions, I have

```ruby
def perform(_, arg_array)
  MyAppTracer.in_span("do_work") do |span|
    span.set_attribute("my_app.queue_name", "awesome_queue")
    
    # Additional business logic code here ...
  end
end
```

Do I need to wrap each call to `MyAppTracer.in_span(...)` in an `if` condition to check the environment variable, like I did in the initializer configuration, or is there a way to refactor the initializer so calls to `MyAppTracer` don't do anything when instrumentation is disabled?

Thank you for any insight on this and please let me know if I can clarify anything further.

**Comment by arielvalentin** [comment]
Thank you for using OTel Ruby! 

Per the specification there is an environment variable that all OTel SDKs must support, which is `OTEL_SDK_DISABLED`:
 
OTEL_SDK_DISABLED | Disable the SDK for all signals | false | Boolean value. If “true”, a no-op SDK implementation will be used for all telemetry signals. Any other value or absence of the variable will have no effect and the SDK will remain enabled. This setting has no effect on propagators configured through the OTEL_PROPAGATORS variable.
-- | -- | -- | --

https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#general-sdk-configuration

It is however not currently implemented by the Ruby SDK: https://github.com/open-telemetry/opentelemetry-ruby/issues/1388

Would you be interested in helping us implement this? 

That would help us become compliant with the latest version of the specification and allow you to disable the SDK entirely without installing any of the instrumentations. 

You may also simulate this behavior in your code by skipping the configure block all together when that envar is set:

```ruby

Rails.application.configure do
  if ENV.fetch("OTEL_SDK_DISABLED", "false") == "false"
      OpenTelemetry::SDK.configure do |c|
        c.service_name = "MY_APP"
        c.use_all
      end
  end
  # This returns a No-Op tracer
  MyAppTracer = OpenTelemetry.tracer_provider.tracer("MY_APP_TRACER")
end
```

**Comment by jbarber** [comment, marked as answer]
Support for OTEL_SDK_DISABLED appears to have been added in https://github.com/open-telemetry/opentelemetry-ruby/pull/1773
